SQLAlchemy with IAM Authentication

JStover95 opened the pull request:
# SQLAlchemy with IAM Authentication

## Summary

This PR overrides the `flask_sqlalchemy.SQLAlchemy` class with a custom `IamSqlAlchemy` class that enables IAM authentication via boto3.

## Changes

- Added a `IamSqlAlchemy` class to `shared/iam_sqlalchemy.py` that overrides Flask SQLAlchemy's `init_app` function.
  - `init_app` now receives two keyword arguments: `use_iam` and `iam_sslmode`.
  - When `use_iam` is true, the `iam_sslmode` ("required" or "preferred") is applied to the connection and an auth token generated by boto3 is used as the password.
- Updated `app.py` and `extensions.py` to use the new `IamSqlAlchemy` class.
  - `use_iam` is set to `False` when the runtime environment is "development" or "testing," and set to `True` otherwise.
- Added unit tests to `tests/tests_shared/test_iam_sqlalchemy.py`.

## How to Test

**Note:** This PR alone does not contain changes required to test in a staging environment.

1. Run unit tests and ensure all pass.
2. Run the application locally and ensure the `/health` endpoint returns successfully.

## Checklist

- [x] New code includes a license statement.
- [x] Any TODOs are made into new issues.
- [x] Code is properly linted (for Markdown, TypeScript, and Python).
- [x] Unit tests are written for any new code.
- [x] All existing unit tests pass.

JStover95 committed:
Add IamSqlAlchemy class for IAM authentication with SQLAlchemy

- Introduce a new `IamSqlAlchemy` class that extends the SQLAlchemy functionality to support IAM authentication. The class includes methods for initializing the app with IAM options and generating authentication tokens using AWS RDS. This enhancement allows for secure database connections using IAM credentials, improving the overall security posture of the application.

JStover95 committed:
Refactor database initialization to support IAM authentication

- Update `register_extensions` in `app.py` to conditionally initialize the database with IAM support based on the environment (development/testing vs. production).
- Replace `SQLAlchemy` with `IamSqlAlchemy` in `extensions.py` to enhance security by integrating IAM authentication for database connections.

JStover95 committed:
Enhance IamSqlAlchemy for improved IAM SSL mode handling

- Introduced `SSLMode` type for `iam_sslmode` parameter in `init_app` method to enforce valid SSL mode values.
- Refactored the constructor to initialize the RDS client directly within the `IamSqlAlchemy` class, ensuring consistent client management.
- Removed redundant client initialization from `init_app` to streamline the database setup process.

JStover95 committed:
Add unit tests for IamSqlAlchemy class

- Introduced a new test file `test_iam_sqlalchemy.py` to validate the functionality of the `IamSqlAlchemy` class.
- Implemented tests for various scenarios including initialization with and without IAM, custom SSL modes, and exception handling during token generation.
- Ensured comprehensive coverage of the `init_app` method and its interaction with the database connection parameters.

JStover95 committed:
Enhance error handling in IamSqlAlchemy class

- Added logging for exceptions during IAM authentication token generation in the `IamSqlAlchemy` class.
- Introduced a logger to capture error messages, improving visibility into failures when creating authentication tokens.

JStover95 committed:
Refactor IAM SQLAlchemy tests to utilize environment-based database URI

- Updated `mock_app` fixture to use `FLASK_DB` environment variable for database URI configuration.
- Introduced `db_uri_params` fixture to extract and provide database connection parameters for tests.
- Modified test cases to assert against parameters derived from the environment, enhancing flexibility and security in test setups.

chris01b commented on shared/iam_sqlalchemy.py:
```diff
@@ -0,0 +1,95 @@
+        ----
+            app: The Flask application instance.
+            use_iam: Whether to use IAM authentication.
+            sslmode: The SSL mode to use.
```
iam_sslmode

chris01b commented on tests/tests_shared/test_iam_sqlalchemy.py:
```diff
@@ -0,0 +1,170 @@
+@pytest.fixture
+def db_uri_params():
+    uri = os.getenv("FLASK_DB")
+    print(uri)
```
Remove

chris01b commented on tests/tests_shared/test_iam_sqlalchemy.py:
```diff
@@ -0,0 +1,170 @@
+        "username": uri.split("@")[0].split(":")[1][2:],
+        "password": uri.split("@")[0].split(":")[2],
+    }
+    print(x)
```
Remove

chris01b commented on tests/tests_shared/test_iam_sqlalchemy.py:
```diff
@@ -0,0 +1,170 @@
+def db_uri_params():
+    uri = os.getenv("FLASK_DB")
+    print(uri)
+    x = {
```
This could work too
```
from urllib.parse import urlparse
parsed = urlparse(uri)
return {
    "hostname": parsed.hostname,
    "port": parsed.port,
    "database": parsed.path.lstrip('/'),
    "username": parsed.username,
    "password": parsed.password,
}
```

chris01b reviewed (dismissed):
Only minor debugging/documentation things

JStover95 committed:
Update parameter name in IamSqlAlchemy class for clarity

- Renamed `sslmode` parameter to `iam_sslmode` in the `init_app` method to enhance clarity and better reflect its purpose in IAM authentication context.

JStover95 committed:
Refactor db_uri_params fixture to utilize urlparse for improved clarity

- Updated the `db_uri_params` fixture in `test_iam_sqlalchemy.py` to use `urlparse` for extracting database connection parameters from the `FLASK_DB` environment variable, enhancing readability and maintainability of the code.
